python/dgl/heterograph.py

```diff
@@ -311,28 +311,618 @@ def __repr__(self):
+
+        >>> import dgl
+        >>> import torch
+
+        **Homogeneous Graphs or Heterogeneous Graphs with A Single Node Type**
+
+        >>> g = dgl.graph((torch.tensor([0, 0, 2]), torch.tensor([0, 1, 2])))
+        >>> g.ndata['hv'] = torch.arange(3).float().reshape(-1, 1)
+        >>> g.edata['he'] = torch.arange(3).float().reshape(-1, 1)
+        >>> g.remove_nodes(torch.tensor([0, 1]))
+        >>> g
+        Graph(num_nodes=1, num_edges=1,
+            ndata_schemes={'hv': Scheme(shape=(1,), dtype=torch.float32)}
+            edata_schemes={'he': Scheme(shape=(1,), dtype=torch.float32)})
+        >>> g.ndata['hv']
+        tensor([[2.]])
+        >>> g.edata['he']
+        tensor([[2.]])
+
+        **Heterogeneous Graphs with Multiple Node Types**
+
+        >>> g = dgl.heterograph({
+        >>>     ('user', 'plays', 'game'): (torch.tensor([0, 1, 1, 2]),
+        >>>                                 torch.tensor([0, 0, 1, 1])),
+        >>>     ('developer', 'develops', 'game'): (torch.tensor([0, 1]),
+        >>>                                         torch.tensor([0, 1]))
+        >>>     })
+        >>> g.remove_nodes(torch.tensor([0, 1]))
+        DGLError: Node type name must be specified
+        if there are more than one node types.
+        >>> g.remove_nodes(torch.tensor([0, 1]), ntype='game')
+        >>> g.num_nodes('user')
+        3
+        >>> g.num_nodes('game')
+        0
+        >>> g.num_edges('plays')
+        0
+
+        See Also
+        --------
+        add_nodes
+        add_edges
+        remove_edges
+        """
+        # TODO(xiangsx): block do not support remove_nodes
+        if ntype is None:
+            if self._graph.number_of_ntypes() != 1:
+                raise DGLError('Node type name must be specified if there are more than one ' \
+                               'node types.')
+
+        nids = utils.prepare_tensor(self, nids, 'u')
+        assert self.number_of_nodes(ntype) > F.as_scalar(F.max(nids, dim=0)), \
+            'The input nids {} is out of the range [0:{})'.format(
+                F.as_scalar(F.max(nids, dim=0)), self.number_of_nodes(ntype))
+
+        ntid = self.get_ntype_id(ntype)
+        nodes = {}
+        for c_ntype in self.ntypes:
+            if self.get_ntype_id(c_ntype) == ntid:
+                old_nids = self.nodes(c_ntype)
+                # trick here, nid_0 is 0 and should be handled
+                old_nids[0] += 1
+                old_nids = F.scatter_row(old_nids, nids, F.full_1d(
+                    len(nids), 0, F.dtype(old_nids), F.context(old_nids)))
+                nodes[c_ntype] = F.tensor(F.nonzero_1d(old_nids), dtype=F.dtype(old_nids))
+            else:
+                nodes[c_ntype] = self.nodes(c_ntype)
+
+        # node_subgraph
+        sub_g = self.subgraph(nodes)
+        self._graph = sub_g._graph
+        self._node_frames = sub_g._node_frames
+        self._edge_frames = sub_g._edge_frames
+
+    def add_selfloop(self):
```

Comment:
For `add_selfloop` and `remove_selfloop`, how about just adding an `etype` argument and letting users select which relation to modify? We raise error for edge type that does not that represent a homograph.

Reply:
Done